src/libs/actions/Report.ts

```diff
@@ -2549,28 +2551,47 @@ function openReportFromDeepLink(url: string) {
     // Navigate to the report after sign-in/sign-up.
     InteractionManager.runAfterInteractions(() => {
         Session.waitForUserSignIn().then(() => {
-            Navigation.waitForProtectedRoutes().then(() => {
-                if (route && Session.isAnonymousUser() && !Session.canAnonymousUserAccessRoute(route)) {
-                    Session.signOutAndRedirectToSignIn(true);
-                    return;
-                }
-
-                // We don't want to navigate to the exitTo route when creating a new workspace from a deep link,
-                // because we already handle creating the optimistic policy and navigating to it in App.setUpPoliciesAndNavigate,
-                // which is already called when AuthScreens mounts.
-                if (new URL(url).searchParams.get('exitTo') === ROUTES.WORKSPACE_NEW) {
-                    return;
-                }
-
-                if (shouldSkipDeepLinkNavigation(route)) {
-                    return;
-                }
-
-                if (isAuthenticated) {
-                    return;
-                }
-
-                Navigation.navigate(route as Route, CONST.NAVIGATION.ACTION_TYPE.PUSH);
+            Onyx.connect({
+                key: ONYXKEYS.NVP_ONBOARDING,
+                callback: (onboarding) => {
+                    Navigation.waitForProtectedRoutes().then(() => {
+                        if (route && Session.isAnonymousUser() && !Session.canAnonymousUserAccessRoute(route)) {
+                            Session.signOutAndRedirectToSignIn(true);
+                            return;
+                        }
+
+                        // We don't want to navigate to the exitTo route when creating a new workspace from a deep link,
+                        // because we already handle creating the optimistic policy and navigating to it in App.setUpPoliciesAndNavigate,
+                        // which is already called when AuthScreens mounts.
+                        if (new URL(url).searchParams.get('exitTo') === ROUTES.WORKSPACE_NEW) {
+                            return;
+                        }
+
+                        if (shouldSkipDeepLinkNavigation(route)) {
+                            return;
+                        }
+
+                        const state = navigationRef.getRootState();
+                        const currentFocusedRoute = findFocusedRoute(state);
+                        const hasCompletedGuidedSetupFlow = Array.isArray(onboarding) || onboarding?.hasCompletedGuidedSetupFlow;
```

Comment:
I think this should be moved to some helper method

Reply:
@mountiny The `hasCompletedGuidedSetupFlow` to reuse it as a selector here? https://github.com/Expensify/App/blob/ed75b6960f688a53bdb8863388ed9a04e0061a55/src/libs/Navigation/NavigationRoot.tsx#L84-L89

Or the whole part below as something like `shouldSkipDeepLinksWhenOnboarding`
```ts
const state = navigationRef.getRootState();
const currentFocusedRoute = findFocusedRoute(state);
const hasCompletedGuidedSetupFlow = Array.isArray(onboarding) || onboarding?.hasCompletedGuidedSetupFlow;

// We need skip deeplinking if the user hasn't completed the guided setup flow.
if (!hasCompletedGuidedSetupFlow) {
  return;
}

if (isOnboardingFlowName(currentFocusedRoute?.name)) {
  Welcome.setOnboardingErrorMessage(Localize.translateLocal('onboarding.purpose.errorBackButton'));
  return;
}
```
Or maybe both? Not sure what you have in mind 😄

Reply:
What you did is what I meant, thanks!